RecoLocalMuon/GEMRecHit/src/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,551 @@
+      // (not making the correction means that the conditions for
+      // forming a cluster are different if we have missing layers -
+      // this could affect events at the boundaries ) 
+      //to be chained, two hits need also to be "close" in phi and eta
+      bool layerRequirementOK = abs(layer_new-layer_old)<5;
+      bool phiRequirementOK = fabs(phi_old-phi_new) < dPhiChainBoxMax;
+      bool etaRequirementOK = fabs(eta_old-eta_new) < dEtaChainBoxMax;
+      
+      if(layerRequirementOK && phiRequirementOK && etaRequirementOK){
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+
+
+
+
+std::vector<ME0Segment> ME0SegAlgoMM::buildSegments(const EnsembleHitContainer& rechits) {
+  std::vector<ME0Segment> me0segs;
+
+  proto_segment.clear();
+  // select hits from the ensemble and sort it 
+  for (auto rh=rechits.begin(); rh!=rechits.end();rh++){
+    proto_segment.push_back(*rh);
+  }
+  if (proto_segment.size() < minHitsPerSegment){
+    return me0segs;
+  }
+  // The actual fit on all hit of the protosegments;
+  this->doSlopesAndChi2();
+  this->fillLocalDirection();
+  AlgebraicSymMatrix protoErrors = this->calculateError();
+  this->flipErrors( protoErrors ); 
+  ME0Segment tmp(proto_segment,protoIntercept, protoDirection, protoErrors,protoChi2);
+  me0segs.push_back(tmp);
+  return me0segs;
+}
+
+//Method doSlopesAndChi2
+// fitSlopes() and  fillChiSquared() are always called one after the other 
+// In fact the code is duplicated in the two functions (as we need 2 loops) - 
+// it is much better to fix that at some point 
+void ME0SegAlgoMM::doSlopesAndChi2(){
+  this->fitSlopes();
+  this->fillChiSquared();
+}
+/* Method fitSlopes
+ *
+ * Perform a Least Square Fit on a segment as per SK algo
+ *
+ */
+void ME0SegAlgoMM::fitSlopes() {
+
+  CLHEP::HepMatrix M(4,4,0);
+  CLHEP::HepVector B(4,0);
+  // In absence of a geometrical construction of the ME0Ensemble take layer 1  
+  const ME0EtaPartition* ens = theEnsemble.first;
+
+  for (auto ih = proto_segment.begin(); ih != proto_segment.end(); ++ih) {
+    const ME0RecHit& hit = (**ih);
+    const ME0EtaPartition* roll  = theEnsemble.second[hit.me0Id()];
+    GlobalPoint gp         = roll->toGlobal(hit.localPosition());
+    // Locat w,r,t, to the first layer;
+    LocalPoint  lp         = ens->toLocal(gp); 
+    // ptc: Local position of hit w.r.t. chamber
+    double u = lp.x();
+    double v = lp.y();
+    double z = lp.z();
+    // ptc: Covariance matrix of local errors 
+    CLHEP::HepMatrix IC(2,2);
+    IC(1,1) = hit.localPositionError().xx();
+    IC(1,2) = hit.localPositionError().xy();
+    IC(2,2) = hit.localPositionError().yy();
+    IC(2,1) = IC(1,2); // since Cov is symmetric
+    // ptc: Invert covariance matrix (and trap if it fails!)
+    int ierr = 0;
+    IC.invert(ierr); // inverts in place
+    if (ierr != 0) {
+      LogDebug("ME0Segment|ME0") << "ME0Segment::fitSlopes: failed to invert covariance matrix=\n" << IC;      
+    }
```

Comment:
@cmsbuild<https://github.com/cmsbuild>, @cvuosalo<https://github.com/cvuosalo> : I was contacting the original authors of the ME0 segment related files for the remaining changes which require the major modifications in the code.  I am restricted this pull request only for the ME0 Rec hits files. All the ME0 Segment related files with the suggested comments and few other modifications will be submitted in a separate pull request.
Please test.
________________________________
From: Carl Vuosalo [notifications@github.com]
Sent: 11 May 2015 18:12
To: cms-sw/cmssw
Cc: Archie Sharma
Subject: Re: [cmssw] adding ME0 for recolocal (#8912)


In RecoLocalMuon/GEMRecHit/src/ME0SegAlgoMM.cc<https://github.com/cms-sw/cmssw/pull/8912#discussion_r30053839>:

> +    // ptc: Local position of hit w.r.t. chamber
> +    double u = lp.x();
> +    double v = lp.y();
> +    double z = lp.z();
> +    // ptc: Covariance matrix of local errors
> +    CLHEP::HepMatrix IC(2,2);
> +    IC(1,1) = hit.localPositionError().xx();
> +    IC(1,2) = hit.localPositionError().xy();
> +    IC(2,2) = hit.localPositionError().yy();
> +    IC(2,1) = IC(1,2); // since Cov is symmetric
> +    // ptc: Invert covariance matrix (and trap if it fails!)
> +    int ierr = 0;
> +    IC.invert(ierr); // inverts in place
> +    if (ierr != 0) {
> +      LogDebug("ME0Segment|ME0") << "ME0Segment::fitSlopes: failed to invert covariance matrix=\n" << IC;
> +    }


Still pending: If the matrix inversion fails, does it make sense to continue on with the calculation? If so, please add a comment explaining what state IC is in after a failed inversion and why it is still useful. If not, please skip the useless calculations after the failed inversion. This issue also appears in the method fillChiSquared.

—
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/8912/files#r30053839>.